### PR TITLE
Persist time entry description edits

### LINKED ISSUE #441: Get first editting up and running
Let's start small and only allow editing of descriptions of time entries. This should work with both running and no longer running entries.

- [x] Edit view opens from the log
- [x] Edit view opens from the running timer
- [x] User can edit the description
- [ ] Changes are persisted

### PR DESCRIPTION
**Pull request by arturs105**
Closes #441 

Time entry description editing is now working properly.

The DTO for now is quite bare, but other stuff will be added to it later.

Squash whenever.

### CODE REVIEW

**Comment by willsb** [comment]
I think that we should stick with the Builder pattern instead. You can use the DTO for, well, data transfer, but you should still use the Builder internally in Foundation to make things consistent.

File: Toggl.Foundation/DataSources/TimeEntriesDataSource.cs [hunk 1]
```diff
@@ -94,5 +101,24 @@ private void safeSetCurrentlyRunningTimeEntry(IDatabaseTimeEntry timeEntry)
             var next = timeEntry == null ? null : TimeEntry.Clean(timeEntry);
             currentTimeEntrySubject.OnNext(next);
         }
+
+        private TimeEntry createUpdatedTimeEntry(IDatabaseTimeEntry timeEntry, EditTimeEntryDto dto)
+            => TimeEntry.Builder.Create(dto.Id)
+                        .SetDescription(dto.Description)
+                        .SetAt(timeEntry.At)
```
Comment (simonrozsival) [inline Toggl.Foundation/DataSources/TimeEntriesDataSource.cs]: We also have to remember the time when the entity was edited for conflict resolution when we pull data from the server.

File: Toggl.Foundation/Models/TimeEntry.cs [hunk 2]
```diff
@@ -56,6 +81,74 @@ public Builder SetBillable(bool billable)
...
+            {
+                if (tagNames == null) return this;
+                TagNames.AddRange(tagNames);
+                return this;
+            }
+
+            public Builder SetTagIds(IEnumerable<long> tagIds)
+            {
+                if (tagIds == null) return this;
+                TagIds.AddRange(tagIds);
```
Comment (willsb) [inline Toggl.Foundation/Models/TimeEntry.cs]: Why name this `Set` if you're effectively `Add`ing? (similar above, on `SetTagNames`)

File: Toggl.Foundation/Models/TimeEntry.cs [hunk 3]
```diff
@@ -56,6 +81,74 @@ public Builder SetBillable(bool billable)
...
+                return this;
+            }
+
+            public Builder SetIsDeleted(bool isDeleted)
+            {
+                IsDeleted = isDeleted;
+                return this;
+            }
+
             private void ensureValidity()
```
Comment (willsb) [inline Toggl.Foundation/Models/TimeEntry.cs]: Hasn't this changed? Considering Simon's comment, maybe we should ensure the `At` date is set (and then fix it elsewhere in the App).

Comment (arturs105) [reply Toggl.Foundation/Models/TimeEntry.cs]: Yeah, but this builder isn't used just to update a time entry. There may be cases where the `At` field doesn't need to be set.

File: Toggl.Foundation/Models/TimeEntry.cs (hunk 2 above)
Comment (arturs105) [reply Toggl.Foundation/Models/TimeEntry.cs]: We could either rename to `AddTagNames` or clear the existing tags before adding the new ones (then the name would be correct).

File: Toggl.Foundation/DataSources/TimeEntriesDataSource.cs [hunk 4]
```diff
@@ -94,5 +101,24 @@ private void safeSetCurrentlyRunningTimeEntry(IDatabaseTimeEntry timeEntry)
...
+                        .SetTaskId(timeEntry.TaskId)
+                        .SetUserId(timeEntry.UserId)
+                        .SetBillable(timeEntry.Billable)
+                        .SetTagNames(timeEntry.TagNames)
+                        .SetIsDeleted(timeEntry.IsDeleted)
+                        .SetProjectId(timeEntry.ProjectId)
+                        .SetCreatedWith(timeEntry.CreatedWith)
+                        .SetWorkspaceId(timeEntry.WorkspaceId)
+                        .SetServerDeletedAt(timeEntry.ServerDeletedAt)
+                        .SetAt(DateTimeOffset.Now)
```
Comment (amulware) [inline Toggl.Foundation/DataSources/TimeEntriesDataSource.cs]: This needs to use the time service.

**Comment by arturs105** [comment]
I also added a test to check that other properties are not changed. Note that such properties do not include the ones that will be editable in the future, but are not implemented currently, such as tags, project, task etc.

File: Toggl.Foundation/Models/TimeEntry.cs (hunk 3 above)
Comment (simonrozsival) [reply Toggl.Foundation/Models/TimeEntry.cs]: Well I don't think there is ever a case when `At` is not set. `DateTimeOffset` is not nullable and the default value is "2001-01-01 12:00:00". But I would prefer if the Builder had a nullable `At` and we validated that it was properly checked. Unless you can think of a case where not having `At` makes sense.

**Comment by amulware** [comment]
There should always be a valid At date, not having it makes no sense.

File: Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs [hunk 5]
```diff
@@ -133,7 +135,10 @@ private void onDeleteCompleted()
 
         private void confirm()
         {
-            throw new NotImplementedException();
+            var dto = new EditTimeEntryDto { Id = Id, Description = Description };
+            confirmDisposable = dataSource.TimeEntries
+                                          .Update(dto)
+                                          .Subscribe((Exception ex) => close(), () => close());
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs]: Should we create issues to decide what happens when this errors?

File: Toggl.Foundation.Tests/DataSources/TimeEntriesDataSourceTests.cs [hunk 6]
```diff
@@ -327,5 +351,77 @@ public void PropagatesErrorIfTimeEntryIsNotInRepository()
...
+        public sealed class TheUpdateMethod : TimeEntryDataSourceTest
+        {
+            private EditTimeEntryDto prepareTest()
+            {
+                var observable = Observable.Return(DatabaseTimeEntry);
+                Repository.GetById(Arg.Is(DatabaseTimeEntry.Id)).Returns(observable);
+                return new EditTimeEntryDto { Id = DatabaseTimeEntry.Id, Description = "New description" };
+            }
+
+            private bool ensurePropertiesNotChanged(IDatabaseTimeEntry timeEntry)
```
Comment (willsb) [inline Toggl.Foundation.Tests/DataSources/TimeEntriesDataSourceTests.cs]: `ensurePropertiesDidNotChange`

File: Toggl.Foundation.Tests/DataSources/TimeEntriesDataSourceTests.cs [hunk 7]
```diff
@@ -327,5 +351,77 @@ public void PropagatesErrorIfTimeEntryIsNotInRepository()
...
+            }
+
+            [Fact]
+            public async ThreadingTask UpdatesTheIsDirtyProperty()
+            {
+                var dto = prepareTest();
+
+                await TimeEntriesSource.Update(dto);
+
+                await Repository.Received().Update(Arg.Is<IDatabaseTimeEntry>(te => te.IsDirty == true));
```
Comment (willsb) [inline Toggl.Foundation.Tests/DataSources/TimeEntriesDataSourceTests.cs]: No  need to compare it with true

File: Toggl.Foundation.Tests/DataSources/TimeEntriesDataSourceTests.cs [hunk 8]
```diff
@@ -327,5 +351,77 @@ public void PropagatesErrorIfTimeEntryIsNotInRepository()
...
+            }
+
+            [Fact]
+            public async ThreadingTask UpdatesTheAtProperty()
+            {
+                var dto = prepareTest();
+
+                await TimeEntriesSource.Update(dto);
+
+                await Repository.Received().Update(Arg.Is<IDatabaseTimeEntry>(te => te.At != DatabaseTimeEntry.At));
```
Comment (willsb) [inline Toggl.Foundation.Tests/DataSources/TimeEntriesDataSourceTests.cs]: This should check that te.At > DatabaseTimeEntry.At

File: Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs (hunk 5 above)
Comment (amulware) [reply Toggl.Foundation.MvvmCross/ViewModels/EditTimeEntryViewModel.cs]: Apart from that it really shouldn't? Later all these sorts of things need to be logged/reported, but for now, I'm not sure what we can really do here.

**Pull request merged**